### DISCUSSION TITLE
Too many data structures with references

### DISCUSSION THREAD

**Discussion by colbyn** [GitoxideLabs/gitoxide#584] [General]
I feel like this library goes a little too far with references in data structures, it makes the API super frustrating at times.

**Comment by colbyn** [comment]
Alternatively (I’ve done this in a drawing app), would be to move functions into traits where a few methods are required and the rest have default implementations, and then you can implement such for data structures with owned and borrowed versions. 

Edit: removed bad examples. I think I went a little too far...

**Comment by Byron** [comment]
> I feel like this library goes a little too far with references in data structures, it makes the API super frustrating at times.

I think it's important to be specific about the kind of frustration to help me understand where the proposed solution, prefer owning over referencing, is coming from.

As for the `Commit` example, `File` is not copyable, and is used in a couple of places. Could it be made to work without references? Most definitely, and maybe that will happen once the commit-graph is actually used by some operations in other crates. I also think `Commit` is missing a few trait implementations, like `Copy` and `Clone`, which will probably be added once it's used or somebody needs them.

Generally, `gitoxide` crates won't take ownership in parameters unless they need it, a statement relevant to the [`Options` and `Context`](https://github.com/Byron/gitoxide/blob/b3691b2a63c99a8f7199f1bf733cc83047b7a882/DEVELOPMENT.md#L126-L130).

As a takeaway, `Repository` is an entry point to all functionality and it's cheaply cloneable. From there, most types will be references it as they rely on its functionality.

**Comment by colbyn** [comment]
@Byron Hey late reply, had to work on different stuff. 

Anyway, I'd actually like to apologize for my tone in my original post, I think I was a little too strong with the criticism or whatnot (idk it just didn’t seem right) and for that I apologize! (Especially since I'm new to this project and I don't yet understand why things are the way they are so it was probably a mistake to suddenly advocate for an overhaul).
